src/storage/store/engine.rs

```diff
@@ -145,75 +156,65 @@ impl ShardEngine {
             }
         }
 
-        let mut merged_user_messages: Vec<message::Message> = vec![];
-        let mut merged_system_messages: Vec<proto::ValidatorMessage> = vec![];
-
-        for msg in &messages {
-            match msg {
-                Message::ValidatorMessage(msg) => {
-                    if let Some(onchain_event) = &msg.on_chain_event {
-                        match self
-                            .onchain_event_store
-                            .merge_onchain_event(onchain_event.clone(), txn_batch)
-                        {
-                            Ok(_) => {
-                                merged_system_messages.push(msg.clone());
-                                // TODO(aditi): Insert into trie.
-                            }
-                            Err(err) => {
-                                error!("Unable to merge onchain event: {}", err)
-                            }
-                        }
-                    }
-                }
-                Message::UserMessage(msg) => {
-                    let data = msg.data.as_ref().ok_or(EngineError::NoMessageData)?;
-                    let msg_type = MessageType::try_from(data.r#type)
-                        .or(Err(EngineError::InvalidMessageType))?;
-                    match msg_type {
-                        MessageType::CastAdd => {
-                            self.cast_store
-                                .merge(&msg, txn_batch)
-                                .map_err(EngineError::new_store_error(msg.hash.clone()))?;
-                            merged_user_messages.push(msg.clone());
-                            self.trie
-                                .insert(&*self.db, txn_batch, vec![msg.hash.clone()])?;
-                        }
-                        unhandled_type => {
-                            return Err(EngineError::UnsupportedMessageType(unhandled_type));
-                        }
-                    }
-                }
-            }
+        let mut snap_txns = self.create_transactions_from_mempool(messages);
+        for snap_txn in &mut snap_txns {
```

Comment:
nit: is snap_txn short for snapchain_txn? I might prefer the longer form as snap_txn implies (on first read) to be introducing a new concept.